stdlib/public/core/StringUTF8Validation.swift

```diff
@@ -0,0 +1,165 @@
+        try guaranteeContinuation()
+        lastValidIndex &+= 4
+      default:
+        fatalError() // TODO: Builtin.unreachable()
+      }
+    }
+    return .success(UTF8ExtraInfo(isASCII: isASCII))
+  } catch {
+    return .error(toBeReplaced: findInvalidRange(buf[lastValidIndex...]))
+  }
+}
+
+internal func repairUTF8(_ input: UnsafeBufferPointer<UInt8>, firstKnownBrokenRange: Range<Int>) -> String {
+  // at any point, the remaining (unprocessed) input is covered by a number of ranges:
+  //
+  //  1. already processed (implicit by being ahead of `goodRange`, may be empty)
+  //  2. `goodRange` (may be empty)
+  //  3. `brokenRange` (never empty)
+  //  4. `remainderRange` (may be empty)
+  //
+  // At the beginning of the process, the `goodRange` starts at the beginning and extends to just before the first
+  // known broken byte. The known broken bytes are covered in the `brokenRange` and everything following that is
+  // the `remainderRange`.
+  // We then copy the `goodRange` into the target buffer and append a UTF8 replacement character. `brokenRange` is
+  // skipped (replaced by the replacement character) and we restart the same process. This time, `goodRange` extends
+  // from the byte after the previous `brokenRange` to the next `brokenRange`.
+  var result = String()
+  let replacementCharacterCount = Unicode.Scalar._replacementCharacter.withUTF8CodeUnits { $0.count }
+  result.reserveCapacity(input.count + 5 * replacementCharacterCount) // extra space for some replacement characters
+
+  var brokenRange: Range<Int> = firstKnownBrokenRange
+  var goodRange: Range<Int> = 0 ..< brokenRange.startIndex
+  repeat {
+    // very likely this capacity reservation does not actually do anything because we reserved space for the entire
+    // input plus up to five replacement characters up front
+    result._guts.reserveCapacity(result._guts.count + goodRange.count + replacementCharacterCount)
+
+    // we can now safely append the next known good bytes and a replacement character
+    result._guts.appendInPlace(UnsafeBufferPointer(rebasing: input[goodRange]),
+                          isASCII: false /* appending replacement character anyway, so let's not bother */)
+    Unicode.Scalar._replacementCharacter.withUTF8CodeUnits {
+      result._guts.appendInPlace($0, isASCII: false)
+    }
+
+    if brokenRange.endIndex == input.endIndex {
+      return result
+    }
+
+    let remainderRange = brokenRange.endIndex ..< input.endIndex
+    let remainder = UnsafeBufferPointer(rebasing: input[remainderRange])
+    switch validateUTF8(remainder) {
+    case .success:
+      result._guts.appendInPlace(remainder, isASCII: false)
+      return result
+    case .error(let newBrokenRangeRebased):
+      brokenRange = (newBrokenRangeRebased.startIndex + remainderRange.startIndex) ..< (newBrokenRangeRebased.endIndex + remainderRange.startIndex)
+      goodRange = remainderRange.startIndex ..< brokenRange.startIndex
+    }
+  } while true
+}
```

Comment:
I think this method can be simplified. You don't need to track so many ranges if the buffer pointer is the loop-variant and gets rebased each iteration.

Something like (pseudo-y):

```
var result = _StringGuts()
var input = input
var brokenRange = firstKnownBrokenRange
while true {
  result._appendInPlace(UBP(rebasing: input[..<brokenRange.lowerBound]))
  U.S._replacementCharacter.withUTF8CodeUnits { result._appendInPlace($0) }
  input = UBP(rebasing: input[brokenRange.upperBound...])
  switch validateUTF8(input) { // Well, assuming validateUTF8 returns success for empty range, we should test this!
    case .success:
      result._appendInPlace(input, isASCII: false)
      return String(result)
    case .error(let newBrokenRange):
      brokenRange = newBrokenRange
  }
}
return String(result)
```

Reply:
Err, with the appropriate reserveCapacity calls in there, of course...

Reply:
- we have an empty  range check
- I came up with something close to that. `brokenRange = newBrokenRange` wouldn't work as we need to "un-rebase" it first 😤 (such a shame that the UBP indices aren't the pointer value which causes this whole rebase-mess)

Reply:
That's why we rebase input, so that brokenRange is always relative to the input range.